### PR TITLE
905: Adjust write exercise for repetition exercise

### PR DESCRIPTION
**Pull request by ztefanie**
### Proposed changes

- [x] Split Write exercise into rendering and logic stuff
- [x] Created abstract class for generic write exercise data and functions
- [x] Moved existing standard and userVocabulary logic into own service
- [ ] ~~Rename StandardWriteExerciseService~~
- [x] Create RepetitionWriteExerciseService

### Side effects
Could break any existing logic of the writing exercise. Should be tested.

### Resolved issues

Fixes: #905

### CODE REVIEW

File: src/routes/write-exercise/services/__tests__/RepetitionWriteExerciseService.spec.ts [hunk 1]
```diff
@@ -0,0 +1,130 @@
...
+    )
+  }
+
+  RepetitionService.moveWordToPreviousSection = jest.fn()
+  RepetitionService.updateWordNodeCard = jest.fn()
+
+  beforeEach(initTestData)
+
+  describe('tryLater', () => {
+    it('should move word to then end and change nothing else in the results nor the nodeCards', () => {
```
Comment (LeandraH) [inline src/routes/write-exercise/services/__tests__/RepetitionWriteExerciseService.spec.ts:60]:
```suggestion
    it('should move word to the end and change nothing else in the results nor the nodeCards', () => {
```

File: src/routes/write-exercise/services/__tests__/StandardWriteExerciseService.spec.tsx [hunk 2]
```diff
@@ -0,0 +1,131 @@
...
+      setCurrentIndex,
+      setIsAnswerSubmitted,
+      setVocabularyItemWithResults
+    )
+  }
+
+  beforeEach(initTestData)
+
+  describe('tryLater', () => {
+    it('should move word to then end and change nothing else in the results', () => {
```
Comment (LeandraH) [inline src/routes/write-exercise/services/__tests__/StandardWriteExerciseService.spec.tsx:56]:
```suggestion
    it('should move word to the end and change nothing else in the results', () => {
```

**Review by LeandraH — COMMENTED** [review]
Tested in emulated iOS, looks great, thank you! A few small issues:

I seem to have a small issue that doesn't happen with all words where
1. the word is ready to be trained again and in Lernstufe 1
2. I tap on "Jetzt wiederholen"
3. without doing the exercise, I tap to end the exercise and confirm that
4. and the word is downgraded to the not-learned bucket.
I would expect the word in that case to not move?

Also, the Solve All cheat in the repetition exercise seems to only solve the current VocabularyItem and then close the exercise. When I have three words to repeat and tap on solve all at the first word, the exercise closes and I still have two words to repeat.

I have also had at least one moment where I had one word to repeat, I did that (manually), and the RepetitionScreen showed that I had no words left to repeat but the footer showed one word to repeat.

**Comment by ztefanie** [comment]
> Tested in emulated iOS, looks great, thank you! A few small issues:
> 
> I seem to have a small issue that doesn't happen with all words where
> 
> 1. the word is ready to be trained again and in Lernstufe 1
> 2. I tap on "Jetzt wiederholen"
> 3. without doing the exercise, I tap to end the exercise and confirm that
> 4. and the word is downgraded to the not-learned bucket.
>    I would expect the word in that case to not move?

I checked this and cannot reproduce. If you do :
1) Start the exercise
2) close the exercise
The progress stays the same, nothing is changed, every word is in the same bucket

If you do)
1) Start the exercise
2) Click "Show solution"
3) Close the exercise
The first word you did "wrong" (=you looked up the solution) is moved to a lower bucket. This is expected behaviour.

Can you please share some more information?

> 
> Also, the Solve All cheat in the repetition exercise seems to only solve the current VocabularyItem and then close the exercise. When I have three words to repeat and tap on solve all at the first word, the exercise closes and I still have two words to repeat.
> 
Good catch, fixed it.

> I have also had at least one moment where I had one word to repeat, I did that (manually), and the RepetitionScreen showed that I had no words left to repeat but the footer showed one word to repeat.

Yes updating of the footer does not yet work correctly, it is part of #906

**Review by LeandraH — APPROVED** [review]
Can't reproduce it anymore 🤷‍♀️

**Review by f1sh1918 — COMMENTED** [review]
Just added this pr to fix building the app with xcode 15.
https://github.com/digitalfabrik/lunes-app/pull/1000

Without i can not test your issue

**Review by f1sh1918 — APPROVED** [review]
Nice :) couldn't find any issues for the writing exercises
Tested on ios
As i can see in #906 the vocabulary of the finished module won't be added to repetition in this issue

**Pull request merged**